graph/src/data/graphql/effort.rs

```diff
@@ -0,0 +1,407 @@
+        }
+    }
+
+    fn add(&mut self, shape_hash: u64, duration: Duration) {
+        let window_size = self.window_size;
+        let bin_size = self.bin_size;
+        let now = Instant::now();
+        self.effort
+            .entry(shape_hash)
+            .or_insert_with(|| MovingStats::new(window_size, bin_size))
+            .add_at(now, duration);
+        self.total.add_at(now, duration);
+    }
+}
+
+// The size of any individual step when we adjust the kill_rate
+const KILL_RATE_STEP: f64 = 0.1;
+
+/// What to log about the state we are currently in
+enum KillStateLogEvent {
+    /// Overload is starting right now
+    Start,
+    /// Overload has been going on for the duration
+    Ongoing(Duration),
+    /// Overload was resolved after duration time
+    Resolved(Duration),
+    /// Don't log anything right now
+    Skip,
+}
+
+struct KillState {
+    // A value between 0 and 1, where 0 means 'respond to all queries'
+    // and 1 means 'do not respond to any queries'
+    kill_rate: f64,
+    // We adjust the `kill_rate` at most every `KILL_RATE_UPDATE_INTERVAL`
+    last_update: Instant,
+    // When the current overload situation started
+    overload_start: Option<Instant>,
+    // Throttle logging while we are overloaded to no more often than
+    // once every 30s
+    last_overload_log: Instant,
+}
+
+impl KillState {
+    fn new() -> Self {
+        let long_ago = Duration::from_secs(86400);
+        Self {
+            kill_rate: 0.0,
+            last_update: Instant::now() - long_ago,
+            overload_start: None,
+            last_overload_log: Instant::now() - long_ago,
+        }
+    }
+
+    fn log_event(&mut self, now: Instant, overloaded: bool) -> KillStateLogEvent {
+        use KillStateLogEvent::*;
+
+        if let Some(overload_start) = self.overload_start {
+            if !overloaded {
+                self.overload_start = None;
+                Resolved(overload_start.elapsed())
+            } else if now.saturating_duration_since(self.last_overload_log)
+                > Duration::from_secs(30)
+            {
+                self.last_overload_log = now;
+                Ongoing(overload_start.elapsed())
+            } else {
+                Skip
+            }
+        } else if overloaded {
+            self.overload_start = Some(now);
+            self.last_overload_log = now;
+            Start
+        } else {
+            Skip
+        }
+    }
+}
+
+pub struct LoadManager {
+    logger: Logger,
+    effort: QueryEffort,
+    store_wait_stats: PoolWaitStats,
+    /// List of query shapes that have been statically blocked through
+    /// configuration
+    blocked_queries: HashSet<u64>,
+    /// List of query shapes that have caused more than `JAIL_THRESHOLD`
+    /// proportion of the work while the system was overloaded. Currently,
+    /// there is no way for a query to get out of jail other than
+    /// restarting the process
+    jailed_queries: RwLock<HashSet<u64>>,
```

Comment:
The motivation behind it is that this would have ended the load issue we had last week almost immediately by jailing the 2 most expensive queries, which truly were the cause of the issue. When a query lands in jail, some human intervention will be needed: if the query truly is the culprit for high load, it should be configured to be banned; if not, the jail threshold might need to be adjusted (you can turn jailing off completely by setting the threshold to a value > 1)